[css-color] Implement color-layers() (without blend mode support)

nt1m opened the pull request:
#### cd1755d8a2fc5ca111e84a250d28dbdb2cd1a218
<pre>
[css-color] Implement color-layers() (without blend mode support)
<a href="https://bugs.webkit.org/show_bug.cgi?id=277837">https://bugs.webkit.org/show_bug.cgi?id=277837</a>
<a href="https://rdar.apple.com/133499954">rdar://133499954</a>

Reviewed by Sam Weinig.

This implements rendering &amp; syntax for color-layers(&lt;color&gt;#) where the list of colors is ordered from top to bottom (like background layers).

<a href="https://drafts.csswg.org/css-color-6/#color-layers">https://drafts.csswg.org/css-color-6/#color-layers</a>

Preliminary piping for the optional &lt;blend-mode&gt; parameter is also added, though the parsing code is commented out until rendering is implemented.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/color-layers-no-blend-mode-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/color-layers-no-blend-mode-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/color-layers-no-blend-mode.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-invalid-color-layers-function-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-invalid-color-layers-function.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-layers-function-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-layers-function.html: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/color/CSSColorLayersResolver.cpp: Added.
(WebCore::blendSourceOver):
* Source/WebCore/css/color/CSSColorLayersResolver.h: Added.
* Source/WebCore/css/color/CSSColorLayersSerialization.h: Added.
(WebCore::serializationForCSSColorLayers):
* Source/WebCore/css/color/CSSUnresolvedColor.h:
* Source/WebCore/css/color/CSSUnresolvedColorLayers.cpp: Added.
(WebCore::serializationForCSS):
(WebCore::createStyleColor):
(WebCore::createColor):
(WebCore::containsCurrentColor):
(WebCore::containsColorSchemeDependentColor):
* Source/WebCore/css/color/CSSUnresolvedColorLayers.h: Added.
* Source/WebCore/css/color/StyleColor.cpp:
(WebCore::StyleColor::StyleColor):
(WebCore::StyleColor::visit):
(WebCore::StyleColor::copy):
* Source/WebCore/css/color/StyleColor.h:
* Source/WebCore/css/color/StyleColorLayers.cpp: Added.
(WebCore::resolveAbsoluteComponents):
(WebCore::resolveColor):
(WebCore::containsNonAbsoluteColor):
(WebCore::containsCurrentColor):
(WebCore::serializationForCSS):
(WebCore::operator&lt;&lt;):
* Source/WebCore/css/color/StyleColorLayers.h: Added.
(WebCore::operator==):
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
(WebCore::CSSPropertyParserHelpers::consumeColorLayersFunction):
(WebCore::CSSPropertyParserHelpers::consumeAColorFunction):

Canonical link: <a href="https://commits.webkit.org/282094@main">https://commits.webkit.org/282094@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f48c939e13ea1215d34b15ed2185483bff748962

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62055 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41409 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14647 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66035 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12600 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64174 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49095 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12940 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49971 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8698 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65124 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38433 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53739 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30803 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35075 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10975 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11531 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/55150 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56872 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11279 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67763 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/61297 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5998 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11041 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57349 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6024 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53688 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57596 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13792 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4913 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/83060 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37209 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14556 "Found 1 new JSC stress test failure: wasm.yaml/wasm/fuzz/memory.js.wasm-eager-jettison (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38293 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39389 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38038 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
```diff
@@ -560,6 +565,46 @@ static std::optional<CSSUnresolvedColor> consumeColorContrastFunction(CSSParserT
+    auto args = consumeFunction(range);
+
+    // FIXME: Parse blend mode:
+    // auto blendMode = BlendMode::Normal;
```
I would suggest leaving out commented-out code.

darinadler commented on Source/WebCore/css/color/CSSColorLayersResolver.h:
```diff
@@ -0,0 +1,43 @@
+
+#include "CSSPropertyParserConsumer+RawTypes.h"
+#include "Color.h"
+#include <optional>
```
This include doesn’t seem to be needed.

nt1m replied:
There's lots of copy-pasted includes, I probably clean those up, thanks for pointing it out

darinadler commented on Source/WebCore/css/color/CSSColorLayersResolver.h:
```diff
@@ -0,0 +1,43 @@
+
+#pragma once
+
+#include "CSSPropertyParserConsumer+RawTypes.h"
```
This include doesn’t seem to be needed.

darinadler commented on Source/WebCore/css/color/CSSColorLayersResolver.h:
```diff
@@ -0,0 +1,43 @@
+
+struct CSSColorLayersResolver {
+    BlendMode blendMode;
+    Vector<Color> colors;
```
We need to include `<wtf/Vector.h>`.

weinig commented on Source/WebCore/css/color/CSSUnresolvedColorLayers.h:
```diff
@@ -0,0 +1,62 @@
+    bool operator==(const CSSUnresolvedColorLayers&) const = default;
+
+    BlendMode blendMode;
+    Vector<UniqueRef<CSSUnresolvedColor>> colors;
```
I haven't quite worked out how happy `Vector` would be about this, but could this be a `Vector<CSSUnresolvedColor>`? (it may be that Vector requires a fully defined type, in which case this would not work).

nt1m replied:
It needs a fully defined type unfortunately.

darinadler commented on Source/WebCore/css/color/CSSColorLayersSerialization.h:
```diff
@@ -0,0 +1,55 @@
+            builder.append(", "_s);
+        serializationForCSS(builder, color);
+        start = false;
+    }
```
I think @weinig created a better idiom for this kind of thing. Is that right?

weinig replied:
Yup, you can do `builder.append(interleave(colorLayers.colors), [](auto& builder, auto& color) { serializationForCSS(builder, color); }, ", "_s))`;

weinig commented on Source/WebCore/css/color/StyleColorLayers.h:
```diff
@@ -0,0 +1,64 @@
+    bool operator==(const StyleColorLayers&) const = default;
+
+    BlendMode blendMode;
+    Vector<StyleColor> colors;
```
This can likely be a FixedVector, since it should always be created from a previously known size.

nt1m replied:
Just tried changing it, and it's a bit more of a pain due to the lack of helpers on `FixedVector` (no containsIf method, also `Vector::map` doesn't have a variant that maps into a new FixedVector.

nt1m replied:
(Not impossible to solve these things, but a followup is probably more appropriate given the changes needed in WTF)

darinadler commented on Source/WebCore/css/color/CSSColorLayersSerialization.h:
```diff
@@ -0,0 +1,55 @@
+    builder.append("color-layers("_s);
+
+    if (colorLayers.blendMode != BlendMode::Normal)
+        builder.append(nameString(toCSSValueID(colorLayers.blendMode)), ", "_s);
```
This should use the more efficient `nameLiteral`.

weinig replied:
Probably `nameLiteralForSerialization()`, just in case any of the blend modes have weird capitalization.

darinadler commented on Source/WebCore/css/color/CSSColorLayersSerialization.h:
```diff
@@ -0,0 +1,55 @@
+
+namespace WebCore {
+
+// TBD: spec link
```
Don’t understand this comment.

nt1m replied:
Leftover from when this didn't have a spec yet, will insert the spec link here.

weinig reviewed (approved):
Excellent work on this!

darinadler commented on Source/WebCore/css/color/CSSUnresolvedColor.h:
```diff
@@ -29,6 +29,7 @@
 #include "CSSUnresolvedAbsoluteColor.h"
 #include "CSSUnresolvedColorHex.h"
 #include "CSSUnresolvedColorKeyword.h"
+#include "CSSUnresolvedColorLayers.h"
```
Why do we need to include this here?

nt1m replied:
std::variant needs complete types

darinadler commented on Source/WebCore/css/color/CSSUnresolvedColorLayers.h:
```diff
@@ -0,0 +1,62 @@
+
+#include "ColorInterpolationMethod.h"
+#include "StyleColor.h"
+#include <variant>
```
Why are we including this?

darinadler commented on Source/WebCore/css/color/CSSUnresolvedColorLayers.h:
```diff
@@ -0,0 +1,62 @@
+#include "ColorInterpolationMethod.h"
+#include "StyleColor.h"
+#include <variant>
+#include <wtf/Forward.h>
```
I think we need `<wtf/Vector.h>`, not just `<wtf/Forward.h>`.

darinadler commented on Source/WebCore/css/color/CSSUnresolvedColorLayers.h:
```diff
@@ -0,0 +1,62 @@
+#pragma once
+
+#include "ColorInterpolationMethod.h"
+#include "StyleColor.h"
```
I don’t fully understand what includes are needed, but it’s not obvious these are.

darinadler reviewed (commented):
Didn’t finish reviewing, but spotted a few things, so leaving these here for now and I will try to return later and do a review, especially if someone else hasn’t yet.

weinig commented on Source/WebCore/css/color/CSSColorLayersResolver.cpp:
```diff
@@ -0,0 +1,50 @@
+    std::optional<Color> resultingColor;
+
+    for (auto& color : colorLayers.colors) {
+        if (!color.isValid())
```
In what cases can a color be invalid here?

nt1m replied:
None I believe, removed this.

nt1m committed:
[css-color] Implement color-layers() (without blend mode support)
https://bugs.webkit.org/show_bug.cgi?id=277837
rdar://133499954

Reviewed by Sam Weinig.

This implements rendering & syntax for color-layers(<color>#) where the list of colors is ordered from top to bottom (like background layers).

https://drafts.csswg.org/css-color-6/#color-layers

Preliminary piping for the optional <blend-mode> parameter is also added, though the parsing code is commented out until rendering is implemented.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/color-layers-no-blend-mode-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/color-layers-no-blend-mode-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/color-layers-no-blend-mode.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-invalid-color-layers-function-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-invalid-color-layers-function.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-layers-function-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-layers-function.html: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/color/CSSColorLayersResolver.cpp: Added.
(WebCore::blendSourceOver):
* Source/WebCore/css/color/CSSColorLayersResolver.h: Added.
* Source/WebCore/css/color/CSSColorLayersSerialization.h: Added.
(WebCore::serializationForCSSColorLayers):
* Source/WebCore/css/color/CSSUnresolvedColor.h:
* Source/WebCore/css/color/CSSUnresolvedColorLayers.cpp: Added.
(WebCore::serializationForCSS):
(WebCore::createStyleColor):
(WebCore::createColor):
(WebCore::containsCurrentColor):
(WebCore::containsColorSchemeDependentColor):
* Source/WebCore/css/color/CSSUnresolvedColorLayers.h: Added.
* Source/WebCore/css/color/StyleColor.cpp:
(WebCore::StyleColor::StyleColor):
(WebCore::StyleColor::visit):
(WebCore::StyleColor::copy):
* Source/WebCore/css/color/StyleColor.h:
* Source/WebCore/css/color/StyleColorLayers.cpp: Added.
(WebCore::resolveAbsoluteComponents):
(WebCore::resolveColor):
(WebCore::containsNonAbsoluteColor):
(WebCore::containsCurrentColor):
(WebCore::serializationForCSS):
(WebCore::operator<<):
* Source/WebCore/css/color/StyleColorLayers.h: Added.
(WebCore::operator==):
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
(WebCore::CSSPropertyParserHelpers::consumeColorLayersFunction):
(WebCore::CSSPropertyParserHelpers::consumeAColorFunction):

Canonical link: https://commits.webkit.org/282094@main
